src/core/ext/transport/chaotic_good/client/chaotic_good_connector.cc

```diff
@@ -0,0 +1,340 @@
+#include "src/core/lib/promise/race.h"
+#include "src/core/lib/promise/sleep.h"
+#include "src/core/lib/promise/try_join.h"
+#include "src/core/lib/promise/try_seq.h"
+#include "src/core/lib/promise/wait_for_callback.h"
+#include "src/core/lib/resource_quota/arena.h"
+#include "src/core/lib/resource_quota/memory_quota.h"
+#include "src/core/lib/resource_quota/resource_quota.h"
+#include "src/core/lib/slice/slice.h"
+#include "src/core/lib/slice/slice_buffer.h"
+#include "src/core/lib/transport/handshaker.h"
+#include "src/core/lib/transport/metadata_batch.h"
+#include "src/core/lib/transport/promise_endpoint.h"
+#include "src/core/lib/transport/transport.h"
+
+namespace grpc_core {
+namespace chaotic_good {
+using grpc_event_engine::experimental::EventEngine;
+namespace {
+void MaybeNotify(const DebugLocation& location, grpc_closure* notify,
+                 grpc_error_handle error) {
+  if (notify != nullptr) {
+    ExecCtx exec_ctx;
+    ExecCtx::Run(location, notify, error);
+  }
+}
+}  // namespace
+ChaoticGoodConnector::ChaoticGoodConnector()
+    : event_engine_(grpc_event_engine::experimental::GetDefaultEventEngine()),
+      handshake_mgr_(std::make_shared<HandshakeManager>()),
+      data_endpoint_latch_(
+          std::make_shared<Latch<std::shared_ptr<PromiseEndpoint>>>()),
+      wait_for_data_endpoint_callback_(std::make_shared<WaitForCallback>()) {
+  channel_args_ = channel_args_.SetObject(event_engine_);
+  channel_args_ =
+      channel_args_.Set(GRPC_ARG_RESOURCE_QUOTA, ResourceQuota::Default());
+  ee_config_ =
+      grpc_event_engine::experimental::ChannelArgsEndpointConfig(channel_args_);
+}
+ChaoticGoodConnector::~ChaoticGoodConnector() {
+  if (connect_activity_ != nullptr) {
+    connect_activity_.reset();
+  }
+}
+
+auto ChaoticGoodConnector::DataEndpointReadSettingsFrame(
+    std::shared_ptr<ChaoticGoodConnector> self) {
+  GPR_ASSERT(self->data_endpoint_ != nullptr);
+  return TrySeq(
+      self->data_endpoint_->ReadSlice(FrameHeader::kFrameHeaderSize),
+      [self = self](Slice slice) mutable {
+        // Read setting frame;
+        // Parse frame header
+        GPR_ASSERT(self->data_endpoint_ != nullptr);
+        auto frame_header_ =
+            FrameHeader::Parse(reinterpret_cast<const uint8_t*>(
+                GRPC_SLICE_START_PTR(slice.c_slice())));
+        return If(
+            frame_header_.ok(),
+            [frame_header_ = *frame_header_, self = self]() {
+              auto frame_header_length = frame_header_.GetFrameLength();
+              return TrySeq(self->data_endpoint_->Read(frame_header_length),
+                            []() { return absl::OkStatus(); });
+            },
+            [status = frame_header_.status()]() { return status; });
+      });
+}
+
+auto ChaoticGoodConnector::DataEndpointWriteSettingsFrame(
+    std::shared_ptr<ChaoticGoodConnector> self) {
+  GPR_ASSERT(self->data_endpoint_ != nullptr);
+  return TrySeq([self = self]() mutable {
```

Comment:
There's only one step in this seq, so I think it's redundant

Could be rewritten as `return [self]() mutable {....};`